manul/src/session/tokio.rs

```diff
@@ -0,0 +1,169 @@
+//! High-level API for executing sessions in `tokio` tasks.
+
+use alloc::{format, vec::Vec};
+
+use rand_core::CryptoRngCore;
+use tokio::sync::mpsc;
+use tracing::{debug, trace};
+
+use super::{
+    message::Message,
+    session::{CanFinalize, RoundOutcome, Session, SessionParameters},
+    transcript::SessionReport,
+    LocalError,
+};
+use crate::protocol::Protocol;
+
+/// The outgoing message from a local session.
+#[derive(Debug)]
+pub struct MessageOut<SP: SessionParameters> {
+    /// The ID of the session that created the message.
+    ///
+    /// Useful when there are several sessions running on a node, pushing messages into the same channel.
+    pub from: SP::Verifier,
+    /// The ID of the session the message is intended for.
+    pub to: SP::Verifier,
+    /// The message to be sent.
+    ///
+    /// Note that the caller is responsible for encrypting the message and attaching authentication info.
+    pub message: Message<SP::Verifier>,
+}
+
+/// The incoming message from a remote session.
+#[derive(Debug)]
+pub struct MessageIn<SP: SessionParameters> {
+    /// The ID of the session the message originated from.
+    ///
+    /// It is assumed that the message's authentication info has been checked at this point.
+    pub from: SP::Verifier,
+    /// The incoming message.
+    pub message: Message<SP::Verifier>,
+}
+
+/// Executes the session waiting for the messages from the `rx` channel
+/// and pushing outgoing messages into the `tx` channel.
+pub async fn run_session<P, SP>(
+    rng: &mut impl CryptoRngCore,
+    tx: &mpsc::Sender<MessageOut<SP>>,
+    rx: &mut mpsc::Receiver<MessageIn<SP>>,
+    session: Session<P, SP>,
+) -> Result<SessionReport<P, SP>, LocalError>
+where
+    P: Protocol<SP::Verifier>,
+    SP: SessionParameters,
+{
+    let mut session = session;
+    // Some rounds can finalize early and put off sending messages to the next round. Such messages
+    // will be stored here and applied after the messages for this round are sent.
+    let mut cached_messages = Vec::new();
+
+    let key = session.verifier();
```

Comment:
```suggestion
    let verifier = session.verifier();
```
…or something more specific to its role? Maybe just `from`?

Reply:
Renamed to `my_id`, and pre-formatted.